src/jobserver.h

```diff
@@ -0,0 +1,107 @@
+/// (previously --jobserver-fds=<val> in older versions of Make) in the MAKEFLAGS
+/// environment variable and creating the respective file descriptors or objects.
+/// On posix systems, <val> is 'fifo:<name>' or '<read_fd>,<write_fd>' for pipes.
+/// On windows systems, <val> is the name of the semaphore.
+///
+/// The class parses the MAKEFLAGS variable and opens the object handle if needed.
+/// Once enabled, Acquire() must be called to acquire a token from the pool.
+/// If a token is acquired, a new command can be started.
+/// Once the command is completed, Release() must be called to return a token.
+/// Make does not care which order a token is recieved or returned.
+struct Jobserver {
+// TODO: jobserver client support for Windows
+#ifndef _WIN32
+  /// Parse the MAKEFLAGS environment variable to receive the path / FDs / name
+  /// of the token pool, and open the handle to the pool if it is an object.
+  /// If a jobserver argument is found in the MAKEFLAGS environment variable,
+  /// and the handle is successfully opened, later calls to Enable() return true.
+  /// If a jobserver argument is found, but the handle fails to be opened,
+  /// the ninja process is aborted with an error.
+  Jobserver();
+
+  /// Before exiting Ninja, ensure that tokens are returned and handles closed.
+  ~Jobserver();
+#endif
+
+  /// Return true if jobserver functionality is enabled and initialized.
+  bool Enabled() const;
+
+  /// Return current token count or initialization signal if negative.
+  int Tokens() const;
+
+  /// Implementation-specific method to acquire a token from the external pool
+  /// which is called for all tokens but returns early for the first token.
+  /// This method is called every time Ninja needs to start a command process.
+  /// Return true on success (token acquired), and false on failure (no tokens
+  /// available). First call always succeeds. Ninja is aborted on read errors.
+  bool Acquire();
+
+  /// Implementation-specific method to release a token to the external pool
+  /// which is called for all tokens but returns early for the last token.
+  /// Return a previously acquired token to the external token pool.
+  /// It must be called for each successful call to Acquire() after the command
+  /// even if subprocesses fail or in the case of errors causing Ninja to exit.
+  /// Ninja is aborted on write errors, and otherwise calls always succeed.
+  void Release();
+
+  /// Loop through Release() to return all tokens. Called before Ninja exits.
+  void Clear();
+
+private:
+  /// The number of currently acquired tokens, or the jobserver status if negative.
+  /// Used to verify that all acquired tokens have been released before exiting,
+  /// and when the implicit (first) token has been acquired (initialization).
+  /// -1: initialized without a token
+  ///  0: uninitialized or disabled
+  /// +n: number of tokens in use
+  int token_count_ = 0;
```

Comment:
I want to point out this concept and ask whether or not this usage of an int is too unusual or non-standard.
This coincides with the last line of the constructor Jobserver::Jobserver() and the value of capacity in CanRunMore() using the absolute value function.

It's pretty easy to rework this. I just happened to have this idea first, to use the token number in place of Enabled() when there actually are no tokens yet.